sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -304,14 +303,25 @@ def _check_categories(self, X):
                 - None if the feature is not categorical
             None if no feature is categorical.
         """
-        X_is_dataframe = _is_pandas_df(X)
+        X_is_dataframe = hasattr(X, "__dataframe__")
+        if X_is_dataframe:
+            categorical_columns_mask = np.asarray(
+                [
+                    c.dtype[0].name == "CATEGORICAL"
+                    and c.describe_categorical["is_dictionary"]
```

Comment:
ATM in polars, pandas and pyarrow `from_dataframe` raises a `NotImplementedError` for categorical columns for which `is_dictionary` is false; columns produced by pandas and polars always have `is_dictionary` True

Reply:
What is the meaning of `is_dictionary`? Maybe you could add an educational inline comment to explain why expect this to be True for our use case.

Reply:
> columns produced by pandas and polars always have is_dictionary True

What about pyarrow?

Reply:
> What about pyarrow?

AFAICT for pyarrow as well `"is_dictionary"` is always `True`, both in the columns it [produces](https://github.com/apache/arrow/blob/76f5976147883fa42dcfb6c775073cce3eaf3606/python/pyarrow/interchange/column.py#L359) and [accepts](https://github.com/apache/arrow/blob/76f5976147883fa42dcfb6c775073cce3eaf3606/python/pyarrow/interchange/from_dataframe.py#L288)

Reply:
the protocol's [description](https://data-apis.org/dataframe-protocol/latest/API.html) for `is_dictionary` is:

> whether a mapping of categorical values to other objects exists

when `is_dictionary` is `True`, the categorical column's attribute `describe_categorical`, which is a dict, has a value for "categories", which is a Column providing an implicit mapping between mapping from int (the indices) to categories

```python
>>> import pandas as pd
>>> df = pd.DataFrame({"A": pd.Series(["b", "a"], dtype="category")})
>>> df["A"].cat.categories
Index(['a', 'b'], dtype='object')
>>> int_col = df.__dataframe__().get_column(0)
>>> int_col.describe_categorical['is_dictionary']
True
>>> int_col.describe_categorical['categories']._col
0    a
1    b
dtype: object
```

so in that case (which seems to be always for pandas, polars and pyarrow) we can read the categories from `describe_categorical["categories"]`.

I suppose when `is_dictionary` is False the category values are stored directly in the column (and `describe_categorical["categories"]` is None), but I don't have an example.
In that case we could get the categories by loading the column, filtering nulls and getting unique values.
That is what is done for non-dataframe inputs, but the code that does it assumes the values are numbers so we may need to adapt it.

that is my impression but I may be misunderstanding the interchange protocol 😅 as I have no experience with it

Reply:
> so in that case (which seems to be always for pandas, polars and pyarrow) we can read the categories from describe_categorical["categories"].

I agree with your analysis but unfortunately I did not find an easy way to re-construct the list of unique categories as a numpy array (e.g. using `numpy.from_dlpack` or even a Python list from the public dataframe API only.

Reply:
I couldn't either (without using pyarrow). Then I guess we should give up on using the interchage protocol to get the categories, and use it only to identify categorical columns. We can then get the categories by loading the column from the original dataframe into a numpy array, as is done in `main`. In that case we don't care about the value of `"is_dictionary"`.
(As I mentioned I have the impression the easiest way to do it is to let the OrdinalEncoder find the categories.)

Reply:
I agree. Then we can later improve the implementation of category introspection in `OrdinalEncoder` a separate, independent.